Add layer to call oAuth endpoints

ArendPeter opened the pull request:
Cognito Demo
https://www.loom.com/share/13650a9a816e4c5492a044217f2c6453

Keycloak Demo
https://www.loom.com/share/e65edbb12ef64fb3af628e83e1f9bf36

This PR adds the logic to call oAuth endpoints, and in the video, I should how this would work with cognito as an example

The user is logged in when they have access tokens stored in their cookies. Here's how they get those cookies

## Using Authorization Code on first login
1. Users hits login, which opens the oAuth providers login endpoint
2. User completes login, and oAuth provider redirects back to our application, along with a code
3. Our application uses the code to call token api implemented on our backend
4. Our backend retrieves tokens from the oAuth provider using the code along with a secret client id

## Using Refresh Token
1. User's previous access token expires, but they still have a refresh token
2. Our application uses the refresh token to call token api implemented on our backend
3. Our backend retrieves tokens from the oAuth provider using the refresh token along with a secret client id

ArendPeter commented:
Current PR status:
1. Update PR to replace cognito w/ keycloak (we've already got a keycloak server running, it should just be a matter of updating the endpoints)
2. We want to make sure we're doing this right, so I'm going to find a auth expert on codementor to give this an additional review

mikefranze commented:
Looks like it failed in the build process, Could not find a declaration file for modules "cors" or "cookie-parser"

ArendPeter commented:
> Looks like it failed in the build process, Could not find a declaration file for modules "cors" or "cookie-parser"

I rebased the PR branch, and ran a fresh install, but it looks like the deployment still fails

I think I've got the cors/cookie-parser references correct in my package.json / package-lock.json?

```
Arend Peter@APC-PRO7 MINGW64 /c/w/star/star-server/backend (oauth-dev)
$ cat package.json | grep co
    "cookie-parser": "^1.4.6",
    "cors": "^2.8.5",
    "@types/cookie-parser": "^1.4.2",
    "@types/cors": "^2.8.12",

Arend Peter@APC-PRO7 MINGW64 /c/w/star/star-server/backend (oauth-dev)
$ cat package-lock.json | grep cors
    "@types/cors": {
      "resolved": "https://registry.npmjs.org/@types/cors/-/cors-2.8.12.tgz",
    "cors": {
      "resolved": "https://registry.npmjs.org/cors/-/cors-2.8.5.tgz",

Arend Peter@APC-PRO7 MINGW64 /c/w/star/star-server/backend (oauth-dev)
$ cat package-lock.json | grep cookie-parser
    "@types/cookie-parser": {
      "resolved": "https://registry.npmjs.org/@types/cookie-parser/-/cookie-parser-1.4.2.tgz",
    "cookie-parser": {
      "resolved": "https://registry.npmjs.org/cookie-parser/-/cookie-parser-1.4.6.tgz",
```

Could you double check if it's a runtime or build error? I'm curious if it fails when installing cors, or if it fails when the code imports cors

mikefranze commented:
Here's what I get when running npm start in the heroku console

> star-server@1.0.0 start /app
> cd backend && npm install && npm start

npm WARN backend@1.0.0 No description
npm WARN backend@1.0.0 No repository field.

added 110 packages from 180 contributors and audited 230 packages in 4.295s

11 packages are looking for funding
  run `npm fund` for details

found 1 high severity vulnerability
  run `npm audit fix` to fix them, or `npm audit` for details

> backend@1.0.0 start /app/backend
> npm run-script build && node ./build/backend/src/index.js


> backend@1.0.0 build /app/backend
> tsc --project ./

src/index.ts:10:18 - error TS7016: Could not find a declaration file for module 'cors'. '/app/backend/node_modules/cors/lib/index.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/cors` if it exists or add a new declaration (.d.ts) file containing `declare module 'cors';`

10 import cors from 'cors';
                    ~~~~~~

src/index.ts:11:26 - error TS7016: Could not find a declaration file for module 'cookie-parser'. '/app/backend/node_modules/cookie-parser/index.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/cookie-parser` if it exists or add a new declaration (.d.ts) file containing `declare module 'cookie-parser';`

11 import cookieParser from 'cookie-parser';

mikefranze commented on backend/package.json:
```diff
@@ -18,10 +18,17 @@
     "typescript": "^4.5.2"
   },
   "devDependencies": {
+    "@types/cookie-parser": "^1.4.2",
```
cors and cookie-parser were put in devDependencies which is ignored by heroku. Move to dependencies instead

mikefranze commented:
Heroku deployment doesn't crash but its now just a blank page

mikefranze reviewed (commented):
Would we want to move things like keycloakBaseUrl  and clientSecret into ENV variables?

ArendPeter commented:
> Would we want to move things like keycloakBaseUrl and clientSecret into ENV variables?

Yep, in particular we should plan to have the ``clientSecret`` in an ENV variable. ``keycloakBaseUrl`` is the same across the frontend and backend so that will be publicly accessible regardless

ArendPeter commented:
> Heroku deployment doesn't crash but its now just a blank page

<img width="721" alt="image" src="https://user-images.githubusercontent.com/9289903/150179729-899f0e29-81cd-4bc0-af49-3379bf1c0ea1.png">

Here's what it shows in the console log. Is there more info in the heroku log?

mikefranze commented:
[star-vote-pr-oauth-dev-qsralie-console-1642614988601.txt](https://github.com/Equal-Vote/star-server/files/7899461/star-vote-pr-oauth-dev-qsralie-console-1642614988601.txt)
It seems to start the server without issue, here's the logs when I run the build script

ArendPeter commented:
We're not sure why the deployment is failing

I've broken the PR into several commits, I'm going to push them one by one and see where the deployment starts failing

[ ] Update backend dependencies for oAuth work
[ ] Update front end dependencies for oAuth work
[ ] Add token api to backend
[ ] Update frontend to work with cognito
[ ] Add support for keycloak

ArendPeter commented:
Looks like it breaks when adding the backend token api, the diff is here https://github.com/Equal-Vote/star-server/pull/22/commits/440b668180c863e4d4d849f93c9b6943a742549e

ArendPeter commented:
Updating the path fixed it! I pushed everything else as well, and it's still working

Aside from that, [someone from code mentor](https://www.codementor.io/@siddajmera) reviewed it, and the PR as been rebased so I think we should be good to go

mikefranze commented:
When I try to log in I get "Invalid parameter: redirect_uri'

ArendPeter commented:
> When I try to log in I get "Invalid parameter: redirect_uri'

Ah ok, I updated the client in the Keycloak admin console to allow the PR domain
<img width="921" alt="image" src="https://user-images.githubusercontent.com/9289903/152095844-102ee38b-312c-4d64-819a-f5060f96c1a0.png">

I also added a similar update in the code

This solves the invalid parameter error, but the process still isn't able to complete. There's probably an error in the backend logs, can you check them?

mikefranze commented:
[star-vote-re-oauth-dev-ng9mif5-logs-1643779044218.txt](https://github.com/Equal-Vote/star-server/files/7983755/star-vote-re-oauth-dev-ng9mif5-logs-1643779044218.txt)

I tried to register but got a 404 error after I clicked submit. Then I tried to log in and it redirected me to the home page but with a session state in the URL. Now when I click login nothing happens except for the session state in the URL changing. Here are the logs when I tried logging in a few times.

ArendPeter commented:
ah ok, that's the same as what I'm seeing

Here's what I'd expect to see, based on my localhost backend log

```
Server started on port 5000
{
  grant_type: 'authorization_code',
  client_id: 'star_vote_web',
  redirect_uri: 'http://localhost:3000'
}
GET TOKEN authorization_code
{
  grant_type: 'authorization_code',
  client_id: 'star_vote_web',
  redirect_uri: 'http://localhost:3000'
}
GET TOKEN authorization_code
success!
Error while requesting a token { error: 'invalid_grant', error_description: 'Code not valid' }
Error: ENOENT: no such file or directory, stat 'C:\w\star\star-server\frontend\build\index.html'
Made it to controller
Getting Elections
```

In the log you showed, it looks like it's not receiving the GET TOKEN api call at all, and in the client side log I see that it's making the api call, but it's timing out

```
https://star-vote-re-oauth-dev-ng9mif5.herokuapp.com:5000/API/Token?grant_type=authorization_code&redirect_uri=https://star-vote-re-oauth-dev-ng9mif5.herokuapp.com&code=2bb7a925-1c19-4eb9-8525-4cd358aab4a0.b71b2482-22e9-481f-a0a1-146cd939c72a.b3e05373-23ee-4df1-a95a-1007881001d8
```

<img width="1089" alt="image" src="https://user-images.githubusercontent.com/9289903/152100280-668d24e9-193c-4f1c-a728-9a6303116062.png">

ArendPeter commented:
I updated the front end to make a relative call to the token api instead of an absolute url path :D (I don't remember why I wasn't doing that originally, I knew it was necessary in some cases, but I don't think this was one of them). I was using a hard coded port, maybe that was incorrect? 

Here's the login flow
1. Click login
2. Login on Keycloak
3. Redirect to homepage
4. Refresh the homepage and you'll see your email (we can find a way to fix the refresh later, that's react thing)

mikefranze commented:
That works! I'll merge but we'll want to make a new route for auth stuff next. I think you said we need to do more to authenticate the tokens next? I've seen how others have done it elsewhere so we should meet and discuss it.

mikefranze commented:
Well now I'm getting the invalid uri for the merge

ArendPeter commented:
> Well now I'm getting the invalid uri for the merge

I think it's because I had a trailing / in the client configuration for the valid redirect_uris. I updated it, and it seems to work now
